File: go/guide_test.go
package fiki_test

// The samples in docs/user-guide.md, run.
//
// A guide whose code does not compile is worse than no guide: a reader trusts it, pastes it, and
// loses an hour to an API that moved. This is an external test package on purpose, so it sees only
// what a consumer sees.

import (
	"testing"

	fiki "github.com/bakobo/fiki/go"
)

func TestTheGuidesSamplesRun(t *testing.T) {
	key := fiki.Generate()
	if len(key.AID()) != 44 {
		t.Fatalf("an AID is 44 characters, got %d", len(key.AID()))
	}

	url := "https://api.example.com/things?limit=1"
	body := []byte(`{"hello": "world"}`)
	headers, err := fiki.SignRequest(key, "POST", url, nil, fiki.SignOptions{Body: body})
	if err != nil {
		t.Fatal(err)
	}

	maxAge := int64(300)
	verdict, err := fiki.VerifyRequest("POST", url, headers,
		fiki.VerifyOptions{Body: body, MaxAge: &maxAge})
	if err != nil {
		t.Fatal(err)
	}
	if verdict.AID != key.AID() {
		t.Errorf("AID = %q, want %q", verdict.AID, key.AID())
	}
}

File: js/test/guide.test.js
// The samples in `docs/user-guide.md`, run.
//
// A guide whose code does not run is worse than no guide: a reader trusts it, pastes it, and loses
// an hour to an API that moved. These are the same calls the guide shows, so a rename that breaks
// a reader's copy-paste breaks the suite first.

import assert from 'node:assert/strict';
import { describe, it } from 'node:test';

import { FikiError, Key, errors, signRequest, verifyRequest } from '../src/index.js';

const url = 'https://api.example.com/things?limit=1';
const body = new TextEncoder().encode(JSON.stringify({ hello: 'world' }));

describe("the user guide's samples", () => {
  it('signs and verifies as shown', async () => {
    const key = await Key.generate();
    assert.equal(key.aid.length, 44);

    const headers = await signRequest({ key, method: 'POST', url, body });
    assert.ok(headers['Signature-Input'] && headers.Signature && headers['Content-Digest']);

    const verdict = await verifyRequest({ method: 'POST', url, headers, body, maxAge: 300 });
    assert.equal(verdict.aid, key.aid);

    await verifyRequest({ method: 'POST', url, headers, body, maxAge: null, expectedAid: key.aid });
    await assert.rejects(
      () => verifyRequest({ method: 'POST', url, headers, body: new TextEncoder().encode('x'), maxAge: null }),
      (e) => e instanceof FikiError && e.constructor.name === 'DigestMismatch',
    );
  });

  it('keeps the browser key promises the guide makes', async () => {
    // Non-extractable by default, and seed throws; extractable on request, and seed is 32 bytes.
    const guarded = await Key.generate();
    assert.throws(() => guarded.seed, errors.MalformedKey);
    const portable = await Key.generate({ extractable: true });
    assert.equal(portable.seed.length, 32);
  });
});
